### PR TITLE
Precompile renderer plugin during installation only (for faster start-up time)

### PR DESCRIPTION
**Pull request by horizon-blue**
One of the common complains that I heard about Bayes3D is that it can take a while to load up the renderer for the first time. Why? Because we have to compile the Nvidiffrast CUDA extension at least once at the start of each session.

However, this process is not only slow, but it also poses other issues. For example, George and I have found that loading Bayes3D renderer within Drake crashes because the use of subprocess. 

Given that we are not modifying the source code of this CUDA extension a lot, I'm proposing that we move the extension compilation to `setup.py` instead, so it only needs to happen once during `pip install .`. The compiled extension will then be cached in the filesystem, which can be imported as `bayes3d.rendering.nvdiffrast.nvdiffrast_plugin_gl`.

The main benefit we get from this change right now is the faster start-up time. In addition, if we ever plan to release Bayes3D as a wheel, then the changes here will allow us to include the prebuilt extension as part of the wheel, so users that install released version from PyPI won't need to worry about any of the C++ dependencies at all. (Just like we don't need to worry about building C++ when we install Numpy or JAX).

Suggestions and comments are welcome :).

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -31,3 +31,9 @@ assets/*
 *.pdf
 *.pkl
 .DS_Store
+
+# C++ extensions
+*.so 
+
+# Distribution / packaging
+build/
```
Comment (sritchie) [inline .gitignore:39]: good catches!

File: setup.py [hunk 2]
```diff
@@ -1,10 +1,108 @@
...
+
 import setuptools
+from torch.utils import cpp_extension
+
+NAME = "bayes3d"
+VERSION = "0.0.1"
+HOME_DIR = os.path.dirname(__file__)
+CPP_SRC_DIR = os.path.join(HOME_DIR, "bayes3d", "rendering", "nvdiffrast", "common")
+
+with open(os.path.join(HOME_DIR, "requirements.txt")) as f:
```
Comment (sritchie) [inline setup.py:12]: woohoo, I'm glad you started the process of making `setup.py` more pro!

File: setup.py [hunk 3]
```diff
@@ -1,10 +1,108 @@
...
+    REQUIRED_PACKAGES = f.read().splitlines()
+
+
+# Nvdiffrast compilation setup
```
Comment (sritchie) [inline setup.py:16]: this looks great, but I wonder if there is a way to limit this work to only occur on installation. Are there setup.py tasks where we don't want all of this to get triggered? reading versions or something like that? I'm looking into this now...

**Comment by sritchie** [comment]
@horizon-blue what are you running to test this? `pip install .`, and then is there a short script you run after that to check that things worked and are faster? I'd love a minimal script here... thank you!

File: setup.py (hunk 3 above)
Comment (horizon-blue) [reply setup.py:16]: My understanding is that `setup.py` will only be executed on installation (i.e. when running `pip install .`). Users are not expected to run `setup.py` manually, and this script wouldn't be executed when we `import bayes3d` during run time either. (so the `if __name__ == "__main__"` condition is actually redundant in this file)

**Comment by horizon-blue** [comment]
@sritchie Thanks for the review! Yes, I tried `pip install .`, then ran a few of the Jupyter notebooks to confirm that everything still works correctly. If you're looking for a short script to test this out, this [`demo.py`](https://github.com/probcomp/bayes3d/blob/main/demo.py) should also works. i.e. 

```python
pip install .
python demo.py
```

(if we manage to get a GPU runner for our CI, then this can be a minimum installation test)

I also tried building a wheel with `pip wheel --no-deps .`. Here's how the `.whl` file looks like: [
bayes3d-0.0.1-cp310-cp310-linux_x86_64.whl](https://drive.google.com/file/d/1WK7VnCYtIuwpzrIVhIjvET2Dx323ywOg/view?usp=sharing)

(I was building this for testing on Colab. This wheel is not publishable to PyPI yet because it's not a "manylinux" build.)

Also, I see that we are trying to migrate to `pyproject.toml`, which is a wonderful idea :). In that case, I can update `setup.py` so that it only takes care of the cpp extension, and let `pyproject.toml` handles the rest of metadata ([as suggested by setuptools's official documentation](https://setuptools.pypa.io/en/latest/userguide/ext_modules.html#building-extension-modules)).

**Comment by sritchie** [comment]
@horizon-blue if you want to get the PR working with pyproject.toml that would be amazing! Happy to follow your lead here :)

**Comment by horizon-blue** [comment]
Okay, I just updated this PR with most of the stuffs moved to `pyproject.toml` :).

To  get of sense of the amount of time we could save with this PR, I used a modified version of the [`demo.py`](https://github.com/probcomp/bayes3d/blob/main/demo.py) script:
```python
from time import perf_counter
import bayes3d as b

intrinsics = b.Intrinsics(
    height=100, width=100, fx=50.0, fy=50.0, cx=50.0, cy=50.0, near=0.001, far=6.0
)

start = perf_counter()
b.setup_renderer(intrinsics)

print("Setup time: ", perf_counter() - start)
```

Before the changes in this PR:
```
[E rasterize_gl.cpp:121] OpenGL version reported as 4.6
Increasing frame buffer size to (width, height, depth) = (128, 128, 1024)
Setup time:  28.01909241499743
```

After:
``` 
[E rasterize_gl.cpp:121] OpenGL version reported as 4.6
Increasing frame buffer size to (width, height, depth) = (128, 128, 1024)
Setup time:  4.2166189509989636
```

Also, I noticed that we have `setuptools_scm` under `build_system` but didn't enabled it, so I took the liberty to turned it on (by adding the `[tool.setuptools_scm]` section in `pyproject.toml`).

File: .gitignore [hunk 4]
```diff
@@ -30,3 +30,10 @@ assets/*
 *.pdf
 *.pkl
 .DS_Store
+
+# C++ extensions
+*.so
+
+# Distribution / packaging
```
Comment (sritchie) [inline .gitignore:37]: good call, glad that we're moving toward producing actual releases here!

File: bayes3d/__init__.py [hunk 5]
```diff
@@ -10,3 +11,5 @@
 from .viz import *
 
 RENDERER = None
+
+__all__ = ["colmap", "distributions", "scene_graph", "utils"]
```
Comment (sritchie) [inline bayes3d/__init__.py:15]: nice, that's great.

File: pyproject.toml [hunk 6]
```diff
@@ -70,3 +71,15 @@ quote-style = "double"
...
+
+[tool.setuptools_scm]
+
+
+[tool.setuptools.packages.find]
+include = ["bayes3d"]
+namespaces = false
+
+[tool.setuptools.dynamic]
+dependencies = {file = ["requirements.txt"]}
```
Comment (sritchie) [inline pyproject.toml:84]: nice, and I expect we'll be working on these a bit too.

File: pyproject.toml [hunk 7]
```diff
@@ -70,3 +71,15 @@ quote-style = "double"
 indent-style = "space"
 skip-magic-trailing-comma = false
 line-ending = "auto"
+
+
+[tool.setuptools_scm]
+
+
+[tool.setuptools.packages.find]
+include = ["bayes3d"]
```
Comment (sritchie) [inline pyproject.toml:80]: does this find all subpackages as well? I guess it does thanks to your `__all__` changes?

**Review by sritchie — APPROVED** [review]
Looks great, thank you for doing this!

File: pyproject.toml (hunk 7 above)
Comment (horizon-blue) [reply pyproject.toml:80]: Yup, this setting tells setuptools to include everything under [`bayes3d/bayes3d`](https://github.com/probcomp/bayes3d/tree/main/bayes3d) as part of the package.

Even without having [`__all__` in `bayes3d/__init__.py`](https://github.com/probcomp/bayes3d/blob/597f1f0f2a42cff0ef8f680ebbde8bf8430711d5/bayes3d/__init__.py#L15), it will still discover all subpackages, though :). The reason I added `__all__` there is to hint ruff that we are intentionally importing some of the modules to make them available directly, so that ruff won't complain that they are imported but not used [on this line](https://github.com/probcomp/bayes3d/blob/597f1f0f2a42cff0ef8f680ebbde8bf8430711d5/bayes3d/__init__.py#L5).

Without [`from . import colmap, distributions, scene_graph, utils`](https://github.com/probcomp/bayes3d/blob/597f1f0f2a42cff0ef8f680ebbde8bf8430711d5/bayes3d/__init__.py#L5) in `bayes3d/__init__.py`, the following lines will fail:
```python
import bayes3d as b
b.distributions
```
Output:
```
AttributeError: module 'bayes3d' has no attribute 'distributions'
```
However, `distributions` still exists as part of Bayes3D package. We just need to import it explicitly:
```python
import bayes3d as b
import bayes3d.distributions
b.distributions  # <- this works just fine
```
Output:
```
<module 'bayes3d.distributions' from '/blahblahblah/site-packages/bayes3d/distributions.py'>
```

Since we used to have [`from . import distributions`, etc.](https://github.com/probcomp/bayes3d/pull/42/files#diff-4e6fc4c661b288b53e857ec36751555a40ce477173c7ccd1ff9373f4248a7d0bL10-L13) prior to https://github.com/probcomp/bayes3d/pull/42, and people have been relying on this ([there are many cases where people use `b.distributions` without explicitly importing the submodule](https://github.com/search?q=repo%3Aprobcomp%2Fbayes3d%20b.distributions&type=code)), I just thought it might be a good idea to add those imports back to `bayes3d/__init__.py`.

### TL;DR

1. Yes, this will find all sub-packages, but it's not because of the `__all__` change.
2. The `__all__` change is to make linter happier about `from . import colmap, distributions, scene_graph, utils`
3. The purpose of `from . import colmap, distributions, scene_graph, utils` is to get rid of the need for explicit import and for backward compatibility

Hope this clarifies things a bit @sritchie, and thanks for the review! :D

**Comment by horizon-blue** [comment]
Oh, I forgot to mention this, but if people *already have an editable installation* of Bayes3D and ran into `ModuleNotFoundError: No module named 'bayes3d.rendering.nvdiffrast.nvdiffrast_plugin_gl'` error after pulling the changes from this PR:

Try running `pip install .` again (or `pip install -e .` for editable installation) to build the plugin.

This is because editable installation only loads changes in Python files, but won't read package metadata (e.g.  things in `setup.py` or `pyproject.toml`). Unless there's future changes to the cpp files, there no need to rebuild the plugin, so this step only need to be done once.
